File: 0501-1000/0759_employeeFreeTime.py
"""
# Definition for an Interval.
class Interval:
    def __init__(self, start: int = None, end: int = None):
        self.start = start
        self.end = end
"""

class Solution:
    def employeeFreeTime(self, schedule: '[[Interval]]') -> '[Interval]':
        allIntervals = []
        for intervals in schedule:
            for interval in intervals:
                allIntervals.append(interval)
        allIntervals.sort(key = lambda x: x.start)
        mergedIntervals = [allIntervals[0]]
        start, end = allIntervals[0].start, allIntervals[0].end
        for interval in allIntervals[1:]:
            last = mergedIntervals[-1]
            if interval.start <= last.end:
                start = last.start
                end = max(last.end, interval.end)
                newInterval = Interval(start, end)
                mergedIntervals[-1] = newInterval
            else:
                mergedIntervals.append(interval)
        freeIntervals = []
        for i in range(1, len(mergedIntervals)):
            start = mergedIntervals[i-1].end
            end = mergedIntervals[i].start
            freeTime = Interval(start, end)
            freeIntervals.append(freeTime)
        return freeIntervals

# Alternative solution

class Solution:
    def employeeFreeTime(self, schedule: '[[Interval]]') -> '[Interval]':
        merged_schedule = schedule[0]
        output = []

        for i in range(1, len(schedule)):
            merged_schedule = self.merge_intervals(merged_schedule, schedule[i])

        for i in range(1, len(merged_schedule)):
            output.append(Interval(merged_schedule[i-1].end, merged_schedule[i].start))

        return output


    def merge_intervals(self, first, second):
        result = []
        i = j = 0

        while i < len(first) and j < len(second):
            if first[i].start > second[j].end:
                result.append(second[j])
                j += 1
            
            elif first[i].end < second[j].start:
                result.append(first[i])
                i += 1
            
            elif first[i].end > second[j].end:
                first[i].start = min(first[i].start, second[j].start)
                first[i].end = max(first[i].end, second[j].end)
                j += 1

            else:
                second[j].start = min(first[i].start, second[j].start)
                second[j].end = max(first[i].end, second[j].end)
                i += 1

        while i < len(first):
            result.append(first[i])
            i += 1

        while j < len(second):
            result.append(second[j])
            j += 1   

        return result  

File: 0501-1000/0759_employeeFreeTime.java
/*
// Definition for an Interval.
class Interval {
    public int start;
    public int end;

    public Interval() {}

    public Interval(int _start, int _end) {
        start = _start;
        end = _end;
    }
};
*/

class Solution {
    public List<Interval> employeeFreeTime(List<List<Interval>> schedule) {
        List<Interval> finalWorkInterval = schedule.get(0);
        List<Interval> result = new ArrayList<>();
        
        for(int i = 1; i < schedule.size(); i++) {
            finalWorkInterval = mergeIntervals(finalWorkInterval, schedule.get(i));
        }
        
        for (int i = 1; i < finalWorkInterval.size(); i++) {
            result.add(new Interval(finalWorkInterval.get(i - 1).end, finalWorkInterval.get(i).start));
        }
        
        return result;
    }
    
    public List<Interval> mergeIntervals(List<Interval> baseInterval, List<Interval> interval) {
        List<Interval> finalInterval = new ArrayList<>(baseInterval);
        finalInterval.addAll(interval);
        
        Collections.sort(finalInterval, (a, b) -> Integer.compare(a.start, b.start));
        List<Interval> result = new ArrayList<>();
        result.add(finalInterval.get(0));
        
        for (int i = 1; i < finalInterval.size(); i++) {
            if (finalInterval.get(i).start <= result.get(result.size() - 1).end) {
                Interval currInterval = new Interval(
                    result.get(result.size() - 1).start, 
                    Math.max(result.get(result.size() - 1).end, finalInterval.get(i).end)
                );
                result.set(result.size() - 1, currInterval);
            } else {
                result.add(finalInterval.get(i));
            }
        }
        
        return result;
    }
}
